### PR TITLE
OSSM-8998 Disable istioctl commands with building ldflags

### PR DESCRIPTION
**Pull request by bmangoen**
**Please provide a description of this PR:**

We want to get rid of hardcoded unsupported commands and to have the possibility to disable them at building time.  
Related issue [OSSM-8998](https://issues.redhat.com/browse/OSSM-8998)

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

**Please check any characteristics that apply to this pull request.**

- [x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: istioctl/cmd/root.go [hunk 1]
```diff
@@ -313,3 +356,14 @@ func unsupportedCmd(cmd *cobra.Command, alternative string) *cobra.Command {
 		},
 	}
 }
+
+// disableUnsupportedCmds is used to disable all the commands that are not supported.
```
Comment (bartoszmajsak) [inline istioctl/cmd/root.go:360]: One thing I wonder - is it possible that we would not support a single sub-command among others? IIUC this only works on the top-level command rights. If the aforementioned case would be needed, we might want to rework this part to be recursive:

```golang
func VisitAll(command *cobra.Command, apply func(command *cobra.Command)) {
	for _, child := range command.Commands() {
		VisitAll(child, apply)
	}
	apply(command)
}


// actual call

VisitAll(root, func(c *cobra.Command) {
	if unsupportedCmds[c.Name()] {
		parent := c.Parent()
		parent.Remove(c)
		parent.AddCommand(unsupportedCmd(c, UnsupportedCmdsMsg, buildUnsupportedCmdsAlternativeMsg(UnsupportedCmdsAlternativeMsg)[c.Name()]))
})

```

File: istioctl/cmd/root.go [hunk 2]
```diff
@@ -313,3 +356,14 @@ func unsupportedCmd(cmd *cobra.Command, alternative string) *cobra.Command {
...
+func disableUnsupportedCmds(cmd *cobra.Command) {
+	unsupportedCmds := buildUnsupportedCmds(UnsupportedCmds)
+	for _, c := range cmd.Commands() {
+		if unsupportedCmds[c.Name()] {
+			cmd.RemoveCommand(c)
+			cmd.AddCommand(unsupportedCmd(c, UnsupportedCmdsMsg, buildUnsupportedCmdsAlternativeMsg(UnsupportedCmdsAlternativeMsg)[c.Name()]))
```
Comment (bartoszmajsak) [inline istioctl/cmd/root.go:366]: I don't think we should be re-building the map of alternative messages for each iteration. How about merging both maps together - each unsupported cmd will have either default message or alternative if specified. WDYT?

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
I like that we are making this change to be externally configured rather than hardcoded. This will be less work in forthcoming releases.

I think we could also move all the new  `var`s and `func`s to a separate file (e.g. `ossm_custom_cmd.go`), to further limit merge conflicts.

I shared few suggestions for you to consider.

It would also great to add a few simple unit tests for edge cases before we merge this PR.

File: istioctl/cmd/root.go [hunk 3]
```diff
@@ -65,6 +65,21 @@ const (
...
+	UnsupportedCmds string
+
+	// General message that is printed for every unsupported commands.
+	// Eg: go build ... -ldflags -X 'PACKAGE_NAME/istioctl/cmd.UnsupportedCmdsMsg=command not supported in <this_context>'
+	UnsupportedCmdsMsg string
+
+	// Specific message that gives the alternative ways of the not supported command.
+	// Eg: go build ... -X 'PACKAGE_NAME/istioctl/cmd.UnsupportedCmdsAlternativeMsg=command1=alternative message cmd1,command2=alternative message cmd2'
+	UnsupportedCmdsAlternativeMsg string
+)
```
[the remark this answers is not in the document]
Comment (bmangoen) [reply istioctl/cmd/root.go]: Right, renamed

File: istioctl/cmd/root.go [hunk 4]
```diff
@@ -301,10 +317,37 @@ func seeExperimentalCmd(name string) *cobra.Command {
 	}
 }
 
-// unsupportedCmd is used to set commands that are not supported for OpenShift Service Mesh.
-func unsupportedCmd(cmd *cobra.Command, alternative string) *cobra.Command {
+// buildUnsupportedCmdsAlternativeMsg builds the map containing the unsupported commands alternative messages
+// from the 'UnsupportedCmdsAlternativeMsg' build ldflag.
+func buildUnsupportedCmdsAlternativeMsg(buildExpr string) map[string]string {
+	cmds := strings.Split(buildExpr, ",")
```
[the remark this answers is not in the document]
Comment (bmangoen) [reply istioctl/cmd/root.go]: Done

File: istioctl/cmd/root.go [hunk 5]
```diff
@@ -301,10 +317,37 @@ func seeExperimentalCmd(name string) *cobra.Command {
...
+	unsupportedCmdsAlternativeMsgs := make(map[string]string)
+	for _, c := range cmds {
+		alt := strings.SplitN(c, "=", 2)
```
[the remark this answers is not in the document]
Comment (bmangoen) [reply istioctl/cmd/root.go]: Done

File: istioctl/cmd/root.go (hunk 1 above)
Comment (bmangoen) [reply istioctl/cmd/root.go]: I'm not sure this case could happen. It would also imply to get another format of the build flag, like `cmd.disabledCmds=cmd1.sub-cmd,cmd2`
@dgn WDYT?

File: istioctl/cmd/root.go (hunk 2 above)
Comment (bmangoen) [reply istioctl/cmd/root.go]: After refactoring, the `disabledCommands` struct contains all the `disableCommand` with their `name`, `disabled` and `alternative` fields.

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]

**Comment by dgn** [comment]
> @dgn @bmangoen After discussing it briefly with @jwendell - as we have now pretty general purpose and unobtrusive solution - do you think it would make sense to propose this upstream?

Not sure-- I don't think they're interested in disabling commands? It's small enough that we can keep it around IMO

**Comment by dgn** [comment]
btw this should also include changing our CI jobs to not skip istioctl-related tests anymore

File: istioctl/cmd/ossm_custom_cmd.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	// Eg: go build ... -ldflags -X PACKAGE_NAME/istioctl/cmd.disabledCmds=command1;command2;command3
+	disabledCmds string
+
+	// Optional: General message that is printed for every disabled commands.
+	// Eg: go build ... -ldflags -X 'PACKAGE_NAME/istioctl/cmd.disabledCmdsMsg=command not supported in <this_context>'
+	disabledCmdsMsg string
+
+	// Optional: Specific message that gives the alternative ways of the disabled commands.
+	// Eg: go build ... -X 'PACKAGE_NAME/istioctl/cmd.disabledCmdsAlternativeMsg=command1=alternative message cmd1;command2=alternative message cmd2'
+	disabledCmdsAlternativeMsg string
```
[the remark this answers is not in the document]
Comment (bmangoen) [reply istioctl/cmd/ossm_custom_cmd.go]: After discussing, I renamed the variable to make it more clear

**Review by bmangoen — COMMENTED** [review]
Changes done

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
Thanks for taking my suggestion into consideration. Great to see it's such a small change now :) Last few minor tweaks and let's merge it :)

File: istioctl/cmd/ossm_custom_cmd.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	disabledCmdsExtraInfo string
+)
+
+const (
+	defaultDisabledCmdsMsg = "command is disabled"
+)
+
+type disabledCommand struct {
+	name      string
+	disabled  bool
```
[the remark this answers is not in the document]
Comment (bmangoen) [reply istioctl/cmd/ossm_custom_cmd.go]: `disabled` is now removed

File: istioctl/cmd/ossm_custom_cmd.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+// disableCmds disables all the flagged "disabled" commands.
+func disableCmds(cmd *cobra.Command) {
+	disabledCommands := newDisableCommands(disabledCmds, disabledCmdsExtraInfo)
+	for _, c := range cmd.Commands() {
+		if disabledCommands[c.Name()] != nil && disabledCommands[c.Name()].disabled {
+			cmd.RemoveCommand(c)
+			cmd.AddCommand(disabledCmd(c, disabledCmdsMsg, disabledCommands[c.Name()].extraInfo))
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (bmangoen) [reply istioctl/cmd/ossm_custom_cmd.go]: Good one, I've implemented as you suggested. Thanks

File: istioctl/cmd/ossm_custom_cmd.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+		extraInfo := strings.SplitN(e, "=", 2)
+		if len(extraInfo) == 2 {
+			commands[extraInfo[0]].extraInfo = extraInfo[1]
+		}
+	}
+	return commands
+}
+
+// disabledCmd is used to set and return a command as disabled.
+func disabledCmd(cmd *cobra.Command, message, alternativeMsg string) *cobra.Command {
```
[the remark this answers is not in the document]
Comment (bmangoen) [reply istioctl/cmd/ossm_custom_cmd.go]: Done
